iceoryx_hoofs/source/posix_wrapper/named_semaphore.cpp

```diff
@@ -0,0 +1,201 @@
+
+#include "iceoryx_hoofs/posix_wrapper/named_semaphore.hpp"
+#include "iceoryx_hoofs/cxx/helplets.hpp"
+#include "iceoryx_hoofs/internal/log/hoofs_logging.hpp"
+#include "iceoryx_hoofs/posix_wrapper/posix_call.hpp"
+
+namespace iox
+{
+namespace posix
+{
+static cxx::string<NamedSemaphore::Name_t::capacity() + 1> createNameWithSlash(const NamedSemaphore::Name_t& name)
+{
+    cxx::string<NamedSemaphore::Name_t::capacity() + 1> nameWithSlash = name;
+    nameWithSlash.insert(0, "/", 1);
+    return nameWithSlash;
+}
+
+static cxx::expected<SemaphoreError> unlink(const NamedSemaphore::Name_t& name) noexcept
+{
+    auto result = posixCall(iox_sem_unlink)(createNameWithSlash(name).c_str())
+                      .failureReturnValue(-1)
+                      .ignoreErrnos(ENOENT)
+                      .evaluate();
+    if (result.has_error())
+    {
+        switch (result.get_error().errnum)
+        {
+        case EACCES:
+            LogError() << "You don't have permission to remove the semaphore \"" << name << "\"";
+            return cxx::error<SemaphoreError>(SemaphoreError::PERMISSION_DENIED);
+        default:
+            LogError() << "This should never happen. An unknown error occurred while creating the semaphore \"" << name
+                       << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::UNDEFINED);
+        }
+    }
+    return cxx::success<>();
+}
+
+cxx::expected<SemaphoreError>
+NamedSemaphoreBuilder::create(cxx::optional<NamedSemaphore>& uninitializedSemaphore) noexcept
+{
+    if (!cxx::isValidFileName(m_name))
+    {
+        LogError() << "The name \"" << m_name << "\" is not a valid semaphore name.";
+        return cxx::error<SemaphoreError>(SemaphoreError::INVALID_NAME);
+    }
+
+    if (m_initialValue > IOX_SEM_VALUE_MAX)
+    {
+        LogError() << "The semaphores \"" << m_name << "\" initial value of " << m_initialValue
+                   << " exceeds the maximum semaphore value " << IOX_SEM_VALUE_MAX;
+        return cxx::error<SemaphoreError>(SemaphoreError::SEMAPHORE_OVERFLOW);
+    }
+
+    if (m_openMode == OpenMode::PURGE_AND_CREATE)
+    {
+        auto result = unlink(m_name);
+        if (result.has_error())
+        {
+            return result;
+        }
+    }
+
+    bool hasOwnership = false;
+
+    auto nameWithSlash = createNameWithSlash(m_name);
+
+    /// BEGIN: try to open existing semaphore
+    if (m_openMode == OpenMode::OPEN_OR_CREATE || m_openMode == OpenMode::OPEN_EXISTING)
+    {
+        auto result = posixCall(iox_sem_open)(nameWithSlash.c_str(), 0)
+                          .failureReturnValue(IOX_SEM_FAILED)
+                          .ignoreErrnos(ENOENT, EINVAL)
+                          .evaluate();
+
+        if (result.has_error())
+        {
+            switch (result.get_error().errnum)
+            {
+            case EACCES:
+                LogError() << "Insufficient permissions to open semaphore \"" << m_name << "\".";
```

Comment:
@elBoberido maybe you don't know about this, but single quote is a shell feature to do stuff like `echo 'hello "bob" world'` but not supported in C++. With single quotes you define a char.
So one could write `char bla = '"';` but not `auto blubb = 'hello "bob" world';`

;)

Reply:
You misunderstood me, I meant `"Hello 'bob' world"`

Reply:
If it is alright with you I would stick with `"`. I just like it more since it is a quotation for me and `'` is something else.

But I have no strong feelings about it.